### PR TITLE
backfill config loader + validation pipeline (#684)

### LINKED ISSUE #684: Backfill 3: Config + validation (TOML schema, CLI merge, all 8 validation rules, chunk-boundary expansion)
## Parent PRD

#678

## What to build

- TOML parser for the backfill config using the existing `pelletier/go-toml` dependency. Sections: `[SERVICE]`, `[BACKFILL]`, `[IMMUTABLE_STORAGE.LEDGERS]`, `[IMMUTABLE_STORAGE.EVENTS]`, `[IMMUTABLE_STORAGE.TXHASH_RAW]`, `[IMMUTABLE_STORAGE.TXHASH_INDEX]`, `[BACKFILL.BSB]`, `[LOGGING]`.
- CLI flag merge layer: `--start-ledger`, `--end-ledger`, `--workers`, `--max-retries`, `--verify-recsplit`, `--log-level`, `--log-format`. CLI flags override TOML.
- `Validate(cfg, flags)` implementing all 8 validation rules from #678 "Validation before DAG construction": ledger-range shape, BSB section presence, `--workers` / `--max-retries` sanity, `CHUNKS_PER_TXHASH_INDEX` immutability via a meta-store-key contract (the actual meta-store read comes later in the metastore-facade slice — this slice wires the contract and uses a stub for the first-read/mismatch check), chunk-boundary expansion (widens range outward), BSB-availability probe (stubbed as an interface for now).
- Subcommand flow: load TOML → merge CLI flags → validate → print expanded range + config summary → exit 0 (no DAG yet).

## Acceptance criteria

- [ ] TOML round-trips through parser + MarshalTOML.
- [ ] All 8 validation rules have their own unit tests, exercising pass and fail paths.
- [ ] Chunk-boundary expansion: table-driven tests for boundary values (start falls mid-chunk, end falls mid-chunk, start=2, end=chunk-boundary).
- [ ] `CHUNKS_PER_TXHASH_INDEX` immutability check calls through a meta-store-read interface stubbed in this slice (real impl wired when metastore facade lands).
- [ ] Running the subcommand against a valid config prints expanded range + config summary, exits 0. Against an invalid config, prints the specific error and exits non-zero.
- [ ] No direct `binary.BigEndian` / `binary.LittleEndian` references — all byte encoding goes through `pkg/format`.

## Blocked by

- Blocked by #680

## User stories addressed (from #678)

- 2, 3, 5, 8, 9, 10

## Per-slice TDD session guidance

> Commit `19a917907e836366d90f189dd0ac0190b45eb8e7` on this repo contains a working-but-exploratory reference implementation of the backfill pipeline under `full-history/backfill/`. Before writing any code for this slice:
> 1. Identify which files in the reference commit correspond to this slice. Use `git show 19a9179:full-history/backfill/<file>.go` to read them (do NOT check them out).
> 2. Read the matching section of `full-history/design-docs/03-backfill-workflow.md` on `feature/full-history`.
> 3. For each reference file, make an explicit port / adapt / rewrite / discard call:
>    - **Port** if it is a pure, self-contained module and matches the current design doc.
>    - **Adapt** if the shape is right but field names, package paths, or config keys need to change for the new home under `cmd/`.
>    - **Rewrite** if the file touches an unresolved open question or predates a design-doc change.
>    - **Discard** intermediate planning artifacts and anything under `full-history/design-docs-og/`.
> 4. Aggressively port `*_test.go` files when the behavior under test still matches the current design — tests encode validated behavior and re-deriving them wastes hours.
> 5. Follow TDD: for each ported or rewritten behavior, one test → one implementation → repeat. Do not bulk-port files without a test driving each behavior.

---

## Helper port (absorbed from closed issue #681)

Also port `pkg/geometry` from commit `19a9179` (plus its `*_test.go`). Validation's chunk-boundary expansion consumes it (ChunkFirstLedger / ChunkLastLedger).

### PR DESCRIPTION
**Pull request by karthikiyer56**
> [!IMPORTANT]
> Base is temporarily `karthik/backfill-impl/scaffold-subcommand` (head of [#699](https://github.com/stellar/stellar-rpc/pull/699)), not `feature/full-history`. This branch stacks on [#699](https://github.com/stellar/stellar-rpc/pull/699); I'll re-target the base after [#699](https://github.com/stellar/stellar-rpc/pull/699) merges.

Closes [#684](https://github.com/stellar/stellar-rpc/issues/684).

## Summary

- Port `pkg/geometry` from reference commit `19a9179` — Index-only API; drops the old `Range*` nomenclature per the current design doc.
- Add `backfill/config/` with the UPPER_SNAKE_CASE TOML schema, plus `ParseConfig` + `MarshalTOML` round-trip.
- `Validate` — required fields + path defaults under `DEFAULT_DATA_DIR`.
- `ApplyFlags` — widens `--start-ledger` / `--end-ledger` outward to chunk boundaries; resolves the `0 = GOMAXPROCS` sentinel on `--workers`; merges `--log-level` / `--log-format` overrides.
- `ValidateFlags` + `ValidateAgainstStore` (`CHUNKS_PER_TXHASH_INDEX` immutability) + `ValidateAgainstBSB` (availability probe).
- `Store` and `BSBAvailabilityProbe` interfaces with in-memory / no-op implementations. Real implementations replace these constructor calls as part of [#689](https://github.com/stellar/stellar-rpc/issues/689) and [#688](https://github.com/stellar/stellar-rpc/issues/688).
- Fill in the subcommand body scaffolded by [#699](https://github.com/stellar/stellar-rpc/pull/699): read TOML → run the pipeline → print summary. Adds `--log-level` / `--log-format` flags.

## Unit tests

- `pkg/geometry`
  - Chunk + index boundary math (`ChunkFirstLedger`, `ChunkLastLedger`, `IndexFirstChunk`, `IndexLastChunk`, etc.).
  - Ledger → chunk and chunk → index round-trips.
  - Index contiguity across 5 consecutive indexes.
- `config/config_test.go`
  - Full-schema TOML parse (every section + nested `[BACKFILL.BSB]`).
  - Default path + numeric resolution under `DEFAULT_DATA_DIR`.
  - `MarshalTOML` round-trip.
- `config/flags_test.go`
  - Chunk-boundary widening — 4 cases (both mid-chunk, both aligned, start=2 + end mid-chunk, start mid-chunk + end aligned).
  - `workers=0` → `GOMAXPROCS`; `max-retries` passes through literally (including 0).
  - Logging override merge.
  - `BuildGeometry` threads `CHUNKS_PER_TXHASH_INDEX`.
- `config/validate_test.go`
  - Required-field rejections (missing `DEFAULT_DATA_DIR`, missing `[BACKFILL.BSB]`, empty `BUCKET_PATH`).
  - `ValidateFlags` — start < 2, end ≤ start, workers < 0, max-retries < 0; workers=0 sentinel passes.
  - `ValidateAgainstStore` — absent / match / mismatch branches via mock `Store`; precondition error when `Validate()` was skipped.
  - `ValidateAgainstBSB` — available / insufficient / probe-error branches via mock probe.
- `backfill/cmd_test.go`
  - Flag registration (all 8 flags present).
  - Golden-path summary (valid config exits 0, summary contains resolved fields).
  - 4 failure paths surfaced through `cmd.Execute` (start < 2, end ≤ start, missing `DEFAULT_DATA_DIR`, missing `[BACKFILL.BSB]`).
  - Missing `--config` file.

### CODE REVIEW

File: cmd/stellar-rpc/internal/fullhistory/pkg/geometry/geometry.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+	// This is the fundamental unit of data — one LFS file per chunk.
+	ChunkSize uint32 = 10_000
+
+	// RangeSize is the number of ledgers in each range (10 million).
+	// Each range contains exactly ChunksPerRange chunks.
+	RangeSize uint32 = 10_000_000
+
+	// ChunksPerRange is the number of chunks in each range.
+	// Derived from RangeSize / ChunkSize (10K), but declared as a constant
+	// to avoid depending on lfs.ChunkSize here.
```
[the remark this answers is not in the document]
Comment (karthikiyer56) [reply cmd/stellar-rpc/internal/fullhistory/pkg/geometry/geometry.go]: Resolved by deletion — `ChunksPerRange` and the entire `Range*` nomenclature have been removed in favor of the Index-only API per the current design doc. The old wording in the comment went with it.

File: cmd/stellar-rpc/internal/fullhistory/pkg/geometry/geometry.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+// RangeLastChunk returns the last chunk ID (inclusive) in a range.
+func (g Geometry) RangeLastChunk(rangeID uint32) uint32 {
+	return (rangeID * g.ChunksPerTxHashIndex) + g.ChunksPerTxHashIndex - 1
+}
+
+// --- Chunk-level methods ---
+
+// ChunkToRangeID returns the range ID that contains a given chunk.
+func (g Geometry) ChunkToRangeID(chunkID uint32) uint32 {
+	return chunkID / g.ChunksPerTxHashIndex
```
[the remark this answers is not in the document]
Comment (karthikiyer56) [reply cmd/stellar-rpc/internal/fullhistory/pkg/geometry/geometry.go]: Resolved by deletion. All `Range*` methods are gone (they were a throwback to an earlier design); the `Index*` methods consistently use `ChunksPerTxHashIndex`. The inconsistency you flagged disappears with the old API surface.

[the remark this answers is not in the document]
Comment (karthikiyer56) [reply cmd/stellar-rpc/internal/fullhistory/pkg/geometry/geometry.go]: Resolved by deletion, same as the above. `ChunkToRangeID` is gone; `IndexID(chunkID)` is the consistent replacement and uses `ChunksPerTxHashIndex` by construction.

File: cmd/stellar-rpc/internal/fullhistory/backfill/config/validate.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+func (c *Config) ValidateFlags(flags CLIFlags) error {
+	if flags.StartLedger < geometry.FirstLedger {
+		return fmt.Errorf("--start-ledger (%d) must be >= %d", flags.StartLedger, geometry.FirstLedger)
+	}
+	if flags.EndLedger <= flags.StartLedger {
+		return fmt.Errorf("--end-ledger (%d) must be > --start-ledger (%d)", flags.EndLedger, flags.StartLedger)
+	}
+	if flags.Workers < 1 {
+		return fmt.Errorf("--workers (%d) must be >= 1", flags.Workers)
+	}
```
[the remark this answers is not in the document]
Comment (karthikiyer56) [reply cmd/stellar-rpc/internal/fullhistory/backfill/config/validate.go]: Good catch. Fixed: `ValidateFlags` now rejects `Workers < 0` only. `0` is the documented GOMAXPROCS sentinel (matches both the cobra flag default and the `(0 = GOMAXPROCS)` help text), and `ApplyFlags` resolves it downstream. Added a dedicated test case `workers=0 (GOMAXPROCS sentinel) passes`.

File: cmd/stellar-rpc/internal/fullhistory/backfill/config/config.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+//
+// Non-TOML fields populated by ApplyFlags (see flags.go) carry `toml:"-"`
+// so they round-trip cleanly through MarshalTOML.
+type Config struct {
+	// Service holds the base data directory used as a default root for
+	// any storage paths the operator does not set explicitly.
+	Service ServiceConfig `toml:"SERVICE"`
+
+	// MetaStore is an optional section; when unset, Validate defaults
+	// Path to {DEFAULT_DATA_DIR}/META/ROCKSDB.
```
[the remark this answers is not in the document]
Comment (karthikiyer56) [reply cmd/stellar-rpc/internal/fullhistory/backfill/config/config.go]: Fixed — doc now matches the lowercase `filepath.Join` literal (`meta/rocksdb`).

File: cmd/stellar-rpc/internal/fullhistory/backfill/cmd_test.go [hunk 5]
```diff
@@ -2,35 +2,177 @@ package backfill
...
+	cmd := NewCmd()
+	var out bytes.Buffer
+	cmd.SetOut(&out)
+	cmd.SetErr(&out)
+	cmd.SetArgs([]string{
+		"--config", "/tmp/does-not-exist-" + t.Name() + ".toml",
+		"--start-ledger", "2",
+		"--end-ledger", "10001",
+		"--workers", "1",
+	})
```
[the remark this answers is not in the document]
Comment (karthikiyer56) [reply cmd/stellar-rpc/internal/fullhistory/backfill/cmd_test.go]: Fixed — the path is now `filepath.Join(t.TempDir(), "does-not-exist.toml")`. The directory exists, the file does not, so the read error is deterministic.

File: cmd/stellar-rpc/internal/fullhistory/backfill/cmd.go [hunk 6]
```diff
@@ -4,29 +4,191 @@ package backfill
...
-			_, err := fmt.Fprintln(cmd.OutOrStdout(), "full-history-backfill: not yet implemented")
-			return err
-		},
+		RunE:  runBackfill,
 	}
 	cmd.Flags().String("config", "", "Path to TOML configuration file (required)")
 	cmd.Flags().Uint32("start-ledger", 0, "First ledger to ingest (inclusive, >= 2)")
 	cmd.Flags().Uint32("end-ledger", 0, "Last ledger to ingest (inclusive, > start-ledger)")
 	cmd.Flags().Int("workers", 0, "Concurrent DAG task slots (0 = GOMAXPROCS)")
 	cmd.Flags().Int("max-retries", 3, "Max retries per task before marking failed")
```
[the remark this answers is not in the document]
Comment (karthikiyer56) [reply cmd/stellar-rpc/internal/fullhistory/backfill/cmd.go:28]: Fixed — dropped the `MaxRetries`-zero sentinel in `ApplyFlags`. cobra already defaults the flag to `3`, so omission works naturally; an explicit `--max-retries=0` is now a legitimate "no retries, fail fast" request and we don't silently rewrite it. Test updated to assert the literal passthrough, and `CLIFlags` docstring rewritten to describe the new semantics.

File: cmd/stellar-rpc/internal/fullhistory/backfill/config/validate.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+	return nil
+}
+
+// ValidateAgainstStore enforces CHUNKS_PER_TXHASH_INDEX immutability. The
+// value gets stringified and stored under ChunksPerTxHashIndexMetaKey on
+// first run; subsequent runs must match or the on-disk layout breaks.
+//
+// Pre-condition: Validate has run, so ChunksPerTxHashIndex is populated.
+// Skipping Validate would seed "0" and permanently lock that value.
+func (c *Config) ValidateAgainstStore(store Store) error {
```
[the remark this answers is not in the document]
Comment (karthikiyer56) [reply cmd/stellar-rpc/internal/fullhistory/backfill/config/validate.go:114]: Fixed — `ValidateAgainstStore` now enforces its precondition explicitly. If `ChunksPerTxHashIndex == 0` it returns an error naming the missing `Validate()` call, rather than silently seeding `"0"` into the store. Added a dedicated test case ("precondition violation: Validate not called first") that constructs an un-Validated `Config` and asserts the error path.

File: cmd/stellar-rpc/internal/fullhistory/pkg/geometry/geometry.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+//   - Chunk  — 10_000 ledgers. The atomic unit of ingestion + crash recovery.
+//   - Index  — ChunksPerTxHashIndex chunks (default 1000, = 10M ledgers).
+//     Cadence at which one RecSplit txhash index gets built.
+//
+// Ledger sequences start at FirstLedger (2), so every boundary formula
+// offsets by that anchor.
+package geometry
+
+// FirstLedger is the first ledger of the Stellar network.
+const FirstLedger uint32 = 2
```
Comment (karthikiyer56) [inline cmd/stellar-rpc/internal/fullhistory/pkg/geometry/geometry.go:15]: Two concepts to internalize before reading this file — **Chunk** (fixed 10,000 ledgers, the atomic unit of ingestion + crash recovery) and **Index** (`ChunksPerTxHashIndex` chunks; the cadence at which one RecSplit txhash index gets built). All backfill math in this package is expressible through those two terms. `FirstLedger=2` because Stellar ledger sequences start there — every boundary formula offsets by that anchor.

File: cmd/stellar-rpc/internal/fullhistory/backfill/config/flags.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+	LogFormat      string
+}
+
+// ApplyFlags merges CLI flag values into c: widens the ledger range to
+// chunk boundaries (outward only — never narrows), resolves default
+// values, and overrides TOML logging fields when the operator passed
+// non-empty values.
+//
+// Precondition: Validate has run, so ChunksPerTxHashIndex is populated.
+func (c *Config) ApplyFlags(flags CLIFlags) {
```
Comment (karthikiyer56) [inline cmd/stellar-rpc/internal/fullhistory/backfill/config/flags.go:31]: Chunk-boundary expansion is **outward only**: start widens DOWN to the first ledger of its chunk, end widens UP to the last. Never narrows. This lets operators pass any range (e.g. `--start-ledger 5_000_000 --end-ledger 56_337_842`) without hand-computing chunk alignment — we always cover at least the requested range, possibly slightly more at each end. The `>= FirstLedger` guards are there to prevent a uint32 underflow if a bogus value reaches `ApplyFlags` before `ValidateFlags` runs; `ValidateFlags` is what finally rejects it.

File: cmd/stellar-rpc/internal/fullhistory/backfill/config/validate.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+	"strconv"
+
+	"github.com/stellar/stellar-rpc/cmd/stellar-rpc/internal/fullhistory/pkg/geometry"
+)
+
+// ChunksPerTxHashIndexMetaKey is the Store key under which the first-run
+// value of CHUNKS_PER_TXHASH_INDEX is seeded. Exported so the real
+// metastore facade (#689) uses the exact same literal — the whole point
+// of the immutability check is defeated if two sites disagree.
+const ChunksPerTxHashIndexMetaKey = "config:chunks_per_txhash_index"
```
Comment (karthikiyer56) [inline cmd/stellar-rpc/internal/fullhistory/backfill/config/validate.go:16]: This key is exported on purpose. Once #689 lands the RocksDB-backed metastore facade, that code will read/write the exact same key. If two sites ever disagree on the literal string, the whole immutability check is defeated — a second run would see an 'absent' key and happily seed a new value next to the old one. Keeping the literal in exactly one place is the point.

File: cmd/stellar-rpc/internal/fullhistory/backfill/config/validate.go [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+// ValidateAgainstStore enforces CHUNKS_PER_TXHASH_INDEX immutability. The
+// value gets stringified and stored under ChunksPerTxHashIndexMetaKey on
+// first run; subsequent runs must match or the on-disk layout breaks.
+//
+// Pre-condition: Validate has run, so ChunksPerTxHashIndex is populated.
+// We enforce that explicitly here instead of trusting the caller — a
+// stray first-run invocation without Validate would otherwise seed "0"
+// into the store and permanently lock the layout to that value.
+func (c *Config) ValidateAgainstStore(store Store) error {
```
Comment (karthikiyer56) [inline cmd/stellar-rpc/internal/fullhistory/backfill/config/validate.go:114]: Three branches to verify below:

1. **Absent key** → seed the current value and pass (first run).
2. **Stored == current** → pass (subsequent run, layout unchanged).
3. **Stored != current** → hard abort, with both values named in the error.

The precondition check at the top protects against a caller who forgot to run `Validate()` first. Without it we would silently seed `"0"` into the store on an un-initialized Config and permanently lock the on-disk layout to that bogus value.

File: cmd/stellar-rpc/internal/fullhistory/backfill/config/interfaces.go [hunk 12]
```diff
@@ -0,0 +1,70 @@
+package config
+
+// Store is the key/value contract the CHUNKS_PER_TXHASH_INDEX immutability
+// check runs against. The interface is deliberately narrow (Get + Put
+// only). This slice ships an in-memory implementation; a real
+// RocksDB-backed one will land as part of #689.
+type Store interface {
```
Comment (karthikiyer56) [inline cmd/stellar-rpc/internal/fullhistory/backfill/config/interfaces.go:7]: Both `Store` (Get + Put) and `BSBAvailabilityProbe` (MaxAvailableLedger) are deliberately narrow. This slice ships in-memory / no-op implementations so the subcommand can run the full pipeline end-to-end today. When the real RocksDB-backed store lands in #689 and the real GCS-backed probe in #688, those implementations drop in at the two `cmd.go` call sites — nothing else in this package has to change.

File: cmd/stellar-rpc/internal/fullhistory/backfill/cmd.go [hunk 13]
```diff
@@ -4,29 +4,191 @@ package backfill
...
 		panic(err)
 	}
 	return cmd
 }
+
+// runBackfill is the pipeline: load TOML → validate → merge flags →
+// validate flags → immutability check → BSB availability → summary.
+// Each step wraps its error so operators see the specific rule that
+// failed.
+func runBackfill(cmd *cobra.Command, _ []string) error {
```
Comment (karthikiyer56) [inline cmd/stellar-rpc/internal/fullhistory/backfill/cmd.go:52]: Pipeline order is load-bearing, not cosmetic:

1. `Validate` fills defaults (including `CHUNKS_PER_TXHASH_INDEX=1000` when absent).
2. `ApplyFlags` depends on those defaults — `BuildGeometry` reads `ChunksPerTxHashIndex` to widen the range.
3. `ValidateFlags` sanity-checks the raw CLI numerics.
4. `ValidateAgainstStore` locks the layout-defining `CHUNKS_PER_TXHASH_INDEX` value — requires `Validate` to have run (enforced as a precondition in validate.go).
5. `ValidateAgainstBSB` checks the widened end ledger against BSB — requires `ApplyFlags` to have run.

Re-ordering any step breaks a downstream consumer's precondition.

**Comment by karthikiyer56** [comment]
Closing this PR (and its parent issue #684) as part of the rewrite of PRD #678.

## Why

TOML config parsing is no longer a backfill concern. Under the new framing in PRD #678, the unified `stellar-rpc` service's `func main` parses the TOML config; backfill consumes an already-parsed Go `Config` struct. The 8-rule validation framework collapses:

- `[BACKFILL]` and `[BACKFILL.BSB]` are obsolete — design doc has flat `[BSB]` at the unified-service level.
- There are no per-run CLI flags. `--start-ledger`, `--end-ledger`, `--workers`, `--max-retries`, `--verify-recsplit` all gone.
- No BSB-availability probe — design doc explicitly says "no source probe; source-coverage problems surface at runtime as task failures".
- Chunk-boundary widening is owned by Phase 1 (catchup) in streaming-side code, not by per-call validate.
- `CHUNKS_PER_TXHASH_INDEX` is being renamed to `LEDGERS_PER_TX_INDEX` (cleaner — domain-natural unit).
- Service-startup `validate_config` (immutability check, etc.) is owned by streaming-side code.
- Per-call `validate(range_start, range_end)` inside `run_backfill` is two asserts inline (bounds + ordering) — no slice needed.

## What survives, and where it lands

The `pkg/geometry` port from commit `19a9179` is good work and still wanted. Land it as a tiny standalone PR if you'd like; it'll be consumed by the first slice that needs ledger/chunk/index arithmetic.

Closing without merge.
